### DISCUSSION TITLE
Fingertip taxel coordinates (3D relative to *_tip frame)

### DISCUSSION THREAD

**Discussion by pattacini** [robotology/community#538] [Q&A] [answered]
Below is the request from @maxymczech as made in https://github.com/robotology/icub-tech-support/issues/1170.

>Hello. I have been working with iCub robot model in Gazebo 11 simulator, trying to add artificial skin sensing capabilities. I have installed `icub-main` and `icub-models`, and I have found taxel coordinates for most body parts in folder `/usr/share/iCub/contexts/skinGui/positions/`:
>
>```
>left_arm_mesh.txt
>left_forearm_mesh.txt
>left_forearm_nomesh.txt
>left_forearm_V2.txt
>left_hand_V2_1.txt
>right_arm_mesh.txt
>right_forearm_mesh.txt
>right_forearm_nomesh.txt
>right_forearm_V2.txt
>right_hand_V2_1.txt
>torso.txt
>```
>
>However, the `hand` files contain only coordinates of the taxels on the palm, no fingertips included. I would be really thankful if you pointed me in the right direction here.

**Comment by pattacini** [comment, marked as answer]
Hi @maxymczech

First of all, thanks once again for your ongoing work on [providing iCub with the skin sensors in Gazebo](https://github.com/robotology/gazebo-yarp-plugins/issues/55#issuecomment-889061477).

We don't have yet standardized the information you're seeking about the positions of the taxels in the fingertips.

What we have so far is the positions of the taxels and their normals expressed wrt a root frame, which by the way is not the one provided by the forward kinematics. You can find this information at https://github.com/icub-tech-iit/electronics-boards-public/tree/master/sensor-boards/ftx2.

The meshes of the fingertips (you can peruse in [Meshlab](https://www.meshlab.net/)) can be found at https://github.com/xEnVrE/robots-viz/tree/master/src/RobotsViz/meshes, where fingertips are denoted as **`full_[Left|Right][Thumb|Index|Middle|Ring|Little]3.obj`** and do correspond to the meshes we have in our CAD.

![101935011-ad3a7500-3bde-11eb-892d-fabf9ee4b5b9](https://user-images.githubusercontent.com/3738070/127695187-f8a2ba70-bb22-463a-b47f-8998cd7b6ac4.gif)

As of now, there's no plan to fill up the gap between the fingertip's root frame and the standard [`iCubFinger`](https://robotology.github.io/robotology-documentation/doc/html/classiCub_1_1iKin_1_1iCubFinger.html) end-effector (that should be somehow at the center of the fingertip itself with the z-axis pointing outward).

Anyway, given that the kinematics is known (see the `iKin` class) and it already contains the description of the last rotation axis (where the fingertip's root frame is attached), it should be fairly easy to come up with such a transformation.

@xEnVrE @simeonedussoni please give a look and amend any incorrect information.

cc @traversaro

**Comment by maxymczech** [comment]
@pattacini Thank you very much for the provided information and references to resources!

**Comment by diogofbsilva** [comment]
Hello @maxymczech @pattacini !

Is there any update regarding this issue as I already have been working with the iCub robot in ISR Lisbon and doing some tests with the tactile sensors, primarly of the fingers and when reading data from the skin_events port I noted that I am not having access to information such as normal direction, center of pressure and geometric centre, presumably because the taxels positions of the fingertips are all 0?

If this issue still persists, what are your recommendations so that I can obtain the normal direction of the contact area and also relate each pressure reading with a tactile sensor? Should I visit the iKin class and come with a transformation that allows me to obtain at least the position of each fingertip?

Thank you for your help!
Diogo Silva

**Reply by pattacini** [reply]
Hi @diogofbsilva 

>Is there any update regarding this issue as I already have been working with the iCub robot in ISR Lisbon and doing some tests with the tactile sensors, primarly of the fingers and when reading data from the skin_events port I noted that I am not having access to information such as normal direction, center of pressure and geometric centre, presumably because the taxels positions of the fingertips are all 0?

Nope, there are no updates on this. We're actively working on a new hand; thus, little chances are that we can get back to it.

x-ref pointing to the most recent information we have on fingertips:
- #668

>If this issue still persists, what are your recommendations so that I can obtain the normal direction of the contact area and also relate each pressure reading with a tactile sensor? Should I visit the iKin class and come with a transformation that allows me to obtain at least the position of each fingertip?

You should come up with a rigid transformation linking the root frame, wherein taxels are expressed, with the iCubFinger frames, which are somehow at the center of the tips with the normal pointing outward. This information may be retrieved from the CAD, I suppose.

Done with this, you'll find the taxels normals at https://github.com/icub-tech-iit/electronics-boards-public/blob/master/other-boards/ft2/docs/3Dtaxelpositions.md.

cc @simeonedussoni

**Comment by diogofbsilva** [comment]
Hello @pattacini !

Thank you so much for your help!
I have one extra thing that I would like to ask/confirm, as in the file https://github.com/robotology/icub-main/blob/master/app/skinGui/conf/positions/left_hand_V2_1.txt there are already the positions of the taxels of the palm. My question is: they are expressed in the root frame shown in the mesh (figure below)?

![mesh_hand](https://github.com/robotology/community/assets/102809063/86230078-1739-4965-9b17-a92fd535169e)

The reason I am asking is because looking into the position file it did not make sense to me that this is the root frame (maybe I am confused).
Can you confirm  if in fact they are expressed in this root frame or if not, what is their root frame?

Thank you!!

**Reply by pattacini** [reply]
Hi @diogofbsilva 

I might be wrong, but I'm afraid we don't have similar information for the taxels of the palms.
CC'ing @simeonedussoni who's certainly more knowledgeable than me in this respect.

**Comment by simeonedussoni** [comment]
hi @diogofbsilva 
the taxels positions are expressed in the reference frame of the electronics board, which origins in correspondence with the **X** highlighted in the figure below
![immagine](https://github.com/robotology/community/assets/31098248/e88df512-7505-4bc8-b1d2-c2755b590480)

in the model there should be the precise value of the translation between the two reference frames.

Hope this clarifies

**Reply by pattacini** [reply]
Most likely, the `X` is the arm end-effector that can be retrieved through the forward kinematics (iKin).

**Reply by diogofbsilva** [reply]
Thank you so much for your help @simeonedussoni @pattacini !
It clarified everything I wanted to know! The fact that most likely, X is the arm end-effector used in the forward kinematics can be extremely helpful to me as well!

**Reply by pattacini** [reply]
>The fact that most likely, X is the arm end-effector used in the forward kinematics can be extremely helpful to me as well!

This should be judged carefully though. The CAD of the electronic board can be explored to verify this correspondence.
